ground/src/main/java/com/google/android/ground/persistence/sync/MediaUploadWorker.kt

```diff
@@ -0,0 +1,184 @@
+import java.io.FileNotFoundException
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import timber.log.Timber
+
+/**
+ * A worker that uploads media associated with submissions to the remote storage in the background.
+ *
+ * This worker should only run when the device has a network connection.
+ *
+ * The worker checks the local storage for any submission mutations that are ready for media upload.
+ * If one or more media associated with the submission do not exist locally, the upload fails and is
+ * marked as such. Otherwise, transient failures are considered eligible for retry.
+ */
+@HiltWorker
+class MediaUploadWorker
+@AssistedInject
+constructor(
+  @Assisted context: Context,
+  @Assisted workerParams: WorkerParameters,
+  private val remoteStorageManager: RemoteStorageManager,
+  private val mutationRepository: MutationRepository,
+  private val userMediaRepository: UserMediaRepository,
+) : CoroutineWorker(context, workerParams) {
+
+  private val loiId: String = workerParams.inputData.getString(LOI_ID) ?: ""
+
+  /**
+   * Left-biased combination of two results. Failures propagate. Success will only be returned in
+   * the case that both results are successful.
+   */
+  private fun <T> combineResults(a: kotlin.Result<T>, b: kotlin.Result<T>): kotlin.Result<T> =
+    if (a.isSuccess) b else a
+
+  override suspend fun doWork(): Result = withContext(Dispatchers.IO) { doWorkInternal() }
+
+  /** Performs media uploads for all eligible submission mutations associated with a given LOI. */
+  private suspend fun doWorkInternal(): Result {
+    check(loiId.isNotEmpty()) { "work was queued for an empty location of interest ID" }
+    Timber.d("Starting media upload for LOI: $loiId")
+
+    // Set of mutations eligible for media upload.
+    val mutations =
+      mutationRepository
+        .getMutations(loiId, MutationEntitySyncStatus.MEDIA_UPLOAD_PENDING)
+        .filterIsInstance<SubmissionMutation>()
+        .filter { it.retryCount < Config.MAX_MEDIA_UPLOAD_RETRY_COUNT }
+        .map {
+          it.copy(
+            syncStatus = Mutation.SyncStatus.MEDIA_UPLOAD_IN_PROGRESS,
+            retryCount = it.retryCount + 1
+          )
+        }
+    mutationRepository.updateMutations(mutations) // Mark MEDIA_UPLOAD_IN_PROGRESS
+
+    val mutationStatusesAfterUpload = uploadMediaAndGetStatus(mutations)
+
+    // The worker proceeds to a fixed point as follows:
+    //   If there are any FAILED or PENDING mutations after the worker operation, we retry.
+    //   During the retry, FAILED mutations will not be processed, while PENDING ones will only be
+    //   reattempted if they satisfy the retry count bound. Eventually, either all of the
+    //   mutations for this LOI will have succeeded, failed, or exceeded the retry count.
+    if (
+      mutationStatusesAfterUpload.any {
+        it == Mutation.SyncStatus.FAILED || it == Mutation.SyncStatus.MEDIA_UPLOAD_PENDING
```

Comment:
As the comment described, it won't, since the worker only looks at MEDIA_UPOLOAD_PENDING states. We check for failed states because a FAILED may be caused by a retriable failure (i.e. anything that wasn't FileNotFound Exception).

But I took your suggestion above to use a separate state for retries, which will make this much more obvious and removes the need to check for FAILURE state here.